[Wrangler] Explain bundling & rules (#6164)
cloudflare/cloudflare-docs: content/workers/wrangler/bundling.md

---
pcx_content_type: how-to
title: Bundling
weight: 5
---

# Wrangler's bundling

{{<Aside type="note">}}

> We recommend using Wrangler's inbuilt bundling, but we understand there are cases where you'll need a bit more flexibility. We've built an escape hatch in the form of [Custom Builds](/workers/wrangler/custom-builds/), which let you replace Wrangler's built-in build step with your own.

Reviewer: It's a programming thing: https://wiki.c2.com/?EscapeHatch

{{</Aside>}}

By default, Wrangler bundles your Worker code using [`esbuild`](https://esbuild.github.io/). This means that Wrangler has built in support for importing modules from [npm](https://www.npmjs.com/) defined in your `package.json`. If you'd like to see the exact code that Wrangler will upload to Cloudflare, you can run `wrangler publish --dry-run --outdir dist`, which will show your Worker code post Wrangler's bundling.

> ## Files which won't be bundled

Reviewer 2: ```suggestion
## Files which will not be bundled
```

> The effect of bundling your Worker code is to take multiple modules and bundle them into one. Sometimes, however, that isn't the behaviour you want. There might be modules that _shouldn't_ be inlined directly into the bundle. For instance, bundling a WASM file into your Javascript Worker wouldn't make a whole lot of sense—that's something you want to upload as a separate module that can be imported at runtime. Wrangler supports this out of the box for the following file types:

Reviewer 2: ```suggestion
Bundling your Worker code is taking multiple modules and bundling them into one. Sometimes, that is not the desired behavior. There might be modules that should not be inlined directly into the bundle. For example, instead of bundling a WASM file into your Javascript Worker, you would want to upload the WASM file as a separate module that can be imported at runtime. Wrangler supports this for the following file types:
```

- `.txt`
- `.html`
- `.bin`
- `.wasm`

> See the [rules configuration](/workers/wrangler/configuration/#bundling) to customise these file types.

Reviewer 2: ```suggestion
Refer to [Bundling configuration](/workers/wrangler/configuration/#bundling) to customize these file types.
```

> This means that for an import like the following:

Reviewer 2: ```suggestion
Review the following import:
```

```js
import data from "./example.html"; // Where `example.html` is a file in your local directory
```

> the variable `data` will be a string containing the contents of `example.html`.

Reviewer 2: ```suggestion
In the above import, the variable `data` will be a string containing the contents of `example.html`.
```

> This is also the basis of Wasm support with Wrangler. To use a Wasm module in a Worker developed with Wrangler, add the following to your Worker:

Reviewer 2: ```suggestion
This is also the basis of Wasm support with Wrangler. To use a Wasm module in a Worker developed with Wrangler, add the following to your Worker:
```

```js

> import wasm from "example.wasm"; // Where `example.wasm` is a file in your local directory

Reviewer 3: ```suggestion
import wasm from "./example.wasm"; // Where `example.wasm` is a file in your local directory
```

export default {
  fetch(request) {
    const module = WebAssembly.instantiate(wasm);
  },
};
```

{{<Aside type="warning">}}

> Cloudflare Workers doesn't support `WebAssembly.instantiateStreaming()`

Reviewer 2: ```suggestion
Cloudflare Workers does not support `WebAssembly.instantiateStreaming()`
```

---
pcx_content_type: concept
title: Bundling
weight: 5
---

# Bundling

By default, Wrangler bundles your Worker code using [`esbuild`](https://esbuild.github.io/). This means that Wrangler has built-in support for importing modules from [npm](https://www.npmjs.com/) defined in your `package.json`. To review the exact code that Wrangler will upload to Cloudflare, run `wrangler publish --dry-run --outdir dist`, which will show your Worker code after Wrangler's bundling.

{{<Aside type="note">}}

> We recommend using Wrangler's inbuilt bundling, but we understand there are cases where you will need more flexibility. We have built an escape hatch in the form of [Custom Builds](/workers/wrangler/custom-builds/), which let you run your own build before Wrangler's built-in one.

Reviewer 3: ```suggestion
We recommend using Wrangler's inbuilt bundling, but we understand there are cases where you will need more flexibility. We have built an escape hatch in the form of [Custom Builds](/workers/wrangler/custom-builds/), which lets you run your own build before Wrangler's built-in one.
```

Reviewer 3: idk, maybe not.

{{</Aside>}}

## Files which will not be bundled

> Bundling your Worker code takes multiple modules and bundles them into one. Sometimes, you might have modules that should not be inlined directly into the bundle. For example, instead of bundling a WASM file into your JavaScript Worker, you would want to upload the WASM file as a separate module that can be imported at runtime. Wrangler supports this for the following file types:

Reviewer 3: ```suggestion
Bundling your Worker code takes multiple modules and bundles them into one. Sometimes, you might have modules that should not be inlined directly into the bundle. For example, instead of bundling a Wasm file into your JavaScript Worker, you would want to upload the Wasm file as a separate module that can be imported at runtime. Wrangler supports this for the following file types:
```
